File: src/main/scala/mrtjp/fengine/assemble/ICAssembler.scala
package mrtjp.fengine.assemble

import mrtjp.fengine.simulate.{ICGate, ICRegister}
import mrtjp.fengine.tiles.FETileMap

trait ICAssembler
{
    def allocRegisterID():Int
    def allocRegisterID(id:Int):Int

    def allocGateID():Int
    def allocGateID(id:Int):Int

    def getRemappedRegisterID(id:Int):Int

    def addRemap(oldID:Int, newID:Int):Unit

    def addRegister(id:Int, r:ICRegister):Unit
    def addGate(id:Int, g:ICGate, drivingRegs:collection.Seq[Int], drivenRegs:collection.Seq[Int]):Unit

    def addTileMap(map:FETileMap, remaps:collection.Map[Int, Int]):Unit
    def addFlatMap(flatMap:ICFlatMap, remaps:collection.Map[Int, Int]):Unit

    def getMapIndex:Int

    def result():ICFlatMap
}

object ICAssembler
{
    def newAssembler:ICAssembler = new ICAssemblerImpl
}

File: src/main/java/mrtjp/fengine/api/ICAssembler.java
package mrtjp.fengine.api;

import mrtjp.fengine.simulate.ICGate;
import mrtjp.fengine.simulate.ICRegister;
import mrtjp.fengine.tiles.FETileMap;

import java.util.List;
import java.util.Map;

public interface ICAssembler {

    int allocRegisterID();
    int allocRegisterID(int id);

    int allocGateID();
    int allocGateID(int id);

    int getRemappedRegisterID(int id);

    void addRemap(int oldID, int newID);

    void addRegister(int id, ICRegister r);
    void addGate(int id, ICGate gate, List<Integer> drivingRegs, List<Integer> drivenRegs);

    void addTileMap(FETileMap map, Map<Integer, Integer> remaps);
    void addFlatMap(ICFlatMap flatMap, Map<Integer, Integer> remaps);

    int getMapIndex();

    ICFlatMap result();
}
